src/main/java/com/microsoft/graph/serializer/DefaultSerializer.java

```diff
@@ -118,59 +118,70 @@ public DefaultSerializer(final ILogger logger) {
-                            if(rawJson != null && field != null && rawJson.get(field.getName()) != null && rawJson.get(field.getName()).isJsonObject() 
-                            		&& rawJson.get(field.getName()).getAsJsonObject().get(pair.getKey()).isJsonObject()) {
-                            	childAdditionalDataManager.setAdditionalData(rawJson.get(field.getName()).getAsJsonObject().get(pair.getKey()).getAsJsonObject());
-                            	setChildAdditionalData((IJsonBackedObject) child,rawJson.get(field.getName()).getAsJsonObject().get(pair.getKey()).getAsJsonObject());
+                                // If the item is a valid Graph object, set its additional data
+                                if (child instanceof IJsonBackedObject) {
+                                    final AdditionalDataManager childAdditionalDataManager = ((IJsonBackedObject) child).additionalDataManager();
+                                    final JsonElement fieldElement = rawJson.get(field.getName());
+                                    if(fieldElement != null && fieldElement.isJsonObject() 
+                                            && fieldElement.getAsJsonObject().get(pair.getKey()) != null
+                                            && fieldElement.getAsJsonObject().get(pair.getKey()).isJsonObject()) {
+                                        childAdditionalDataManager.setAdditionalData(fieldElement.getAsJsonObject().get(pair.getKey()).getAsJsonObject());
+                                        setChildAdditionalData((IJsonBackedObject) child,fieldElement.getAsJsonObject().get(pair.getKey()).getAsJsonObject());
+                                    }
+                                }
                             }
                         }
-                    }
-                }
-                // If the object is a list of Graph objects, iterate through elements
-                else if (fieldObject instanceof List && rawJson != null) {
-                    final JsonElement collectionJson = rawJson.get(field.getName());
-                    final List<?> fieldObjectList = (List<?>) fieldObject;
-                    if (collectionJson.isJsonArray() && ((JsonArray)collectionJson).size() == fieldObjectList.size()) {
-                        final JsonArray rawJsonArray = (JsonArray) collectionJson;
-                        for (int i = 0; i < fieldObjectList.size(); i++) {
-                            final Object element = fieldObjectList.get(i);
-                            if (element instanceof IJsonBackedObject) {
-                                final JsonElement elementRawJson = rawJsonArray.get(i);
-                                setChildAdditionalData((IJsonBackedObject) element, elementRawJson.getAsJsonObject());
+                        // If the object is a list of Graph objects, iterate through elements
+                        else if (fieldObject instanceof List) {
```

Comment:
Add a test for `fieldObject instanceof HashMap` using `serializer.deserializeObject(sourcestring, class);`

One test for a homogenous list of objects, and one for a hetergenous list of objects.

Reply:
added a unit test. We currently don't have a case where hashmaps have heterogeneous values and I didn't want to add too much fake code in the test setup.